Mathlib/Analysis/Calculus/FDeriv/Analytic.lean

```diff
@@ -200,3 +200,145 @@ theorem AnalyticOn.iterated_deriv [CompleteSpace F] (h : AnalyticOn 𝕜 f s) (n
 #align analytic_on.iterated_deriv AnalyticOn.iterated_deriv
 
 end deriv
+
+
+section fderiv
+
+variable {p : FormalMultilinearSeries 𝕜 E F} {r : ℝ≥0∞} {n : ℕ}
+
+variable {f : E → F} {x : E} {s : Set E}
+
+/-! The case of continuously polynomial functions. We get the same differentiability
+results as for analytic functions, but without the assumptions that `F` is complete.-/
+
+theorem HasFiniteFPowerSeriesOnBall.differentiableOn
+    (h : HasFiniteFPowerSeriesOnBall f p x n r) : DifferentiableOn 𝕜 f (EMetric.ball x r) :=
+fun _ hy => (h.cPolynomialAt_of_mem hy).analyticAt.differentiableWithinAt
+
+
+theorem HasFiniteFPowerSeriesOnBall.hasFDerivAt (h : HasFiniteFPowerSeriesOnBall f p x n r)
+    {y : E} (hy : (‖y‖₊ : ℝ≥0∞) < r) :
+    HasFDerivAt f (continuousMultilinearCurryFin1 𝕜 E F (p.changeOrigin y 1)) (x + y) :=
+  (h.changeOrigin hy).toHasFPowerSeriesOnBall.hasFPowerSeriesAt.hasFDerivAt
+
+
+theorem HasFiniteFPowerSeriesOnBall.fderiv_eq (h : HasFiniteFPowerSeriesOnBall f p x n r)
+    {y : E} (hy : (‖y‖₊ : ℝ≥0∞) < r) :
+    fderiv 𝕜 f (x + y) = continuousMultilinearCurryFin1 𝕜 E F (p.changeOrigin y 1) :=
+  (h.hasFDerivAt hy).fderiv
+
+
+/-- If a function has a finite power series on a ball, then so does its derivative. -/
+theorem HasFiniteFPowerSeriesOnBall.fderiv (h : HasFiniteFPowerSeriesOnBall f p x (n + 1) r) :
+    HasFiniteFPowerSeriesOnBall (fderiv 𝕜 f)
+      ((continuousMultilinearCurryFin1 𝕜 E F :
+            (E[×1]→L[𝕜] F) →L[𝕜] E →L[𝕜] F).compFormalMultilinearSeries
+        (p.changeOriginSeries 1))
+      x n r := by
```

Comment:
> I know you probably adapted the proof of the existing [HasFPowerSeriesOnBall.fderiv](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FDeriv/Analytic.html#HasFPowerSeriesOnBall.fderiv) (which assumes CompleteSpace), but I wonder why that statement doesn't use [FormalMultilinearSeries.shift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FormalMultilinearSeries.html#FormalMultilinearSeries.shift) ... I think we should look into proving they are equal.

I don't think that they are equal. I tried proving that they are, and the expression used in `HasFPowerSeriesOnBall.fderiv` seems to scramble the variables more, as well as having an extra factor of `n+1` in degree `n`. The first point would not be surprising, as the only thing we know about the two formal multilinear series is that they have the same sum, which does not determine them (unless they are symmetric and we are in characteristic zero). The second point... well, are we sure that `FormalMultilinearSeries.shift` really is the Taylor series of the derivative ? The mathlib file says that it is but doesn't give any explanation or reference. I'll need a pen and paper to investigate more.

> Once we have this, I think the main goal of #9130 can be achieved by identifying `shift` with `linearDeriv` in the case of a FormalMultilinearSeries of a single term, right? We should probably add it the equality to the other file after this is merged.

I would be happy to make the definition of `linearDeriv` look more natural, but I don't understand what you mean here. What is a `FormalMultiliearSeries` of a single term, and how would you adapt the definition of `shift` so that it applies in the context of #9130, where you're working with multilinear maps ? Or do you mean that we should prove that the `FormalMultilinearSeries` of `f.linearDeriv` is equal to the shift of the `FormalMultilinearSeries` of `f` ? (I'm not sure that's true though.) 
By the way, I started rewriting the next step, i.e. the proof that continuous multilinear maps are analytic, [there](https://github.com/smorel394/ExteriorPowers/blob/master/ExteriorPowers/ContinuousMultilinearMapContDiff.lean). The things I am using so far fro #9130 are `domDomRestrict`, `toFormalMultilinearSeries_fixedDegree` (and its lemmas) and `toFormalMultilinearSeries_partialSum`. It totally forgot about `linearDeriv`, I should probably probe that it is indeed the Fréchet derivative...

Reply:
So what should I do with this suggestion ? As I said, `FormalMultilinearSeries.shift` is not equal to the formal power series of the derivative that is currently in mathlib. There are two issues:
(1) Formal power series are not unique, so we could at most expect the two formal power series to coincide on the diagonal. This is not a big issue, as we only care about the behavior on the diagonal, but if we want to use `shift` it means rewriting a pretty big proof.
(2) The normalization: usually Taylor series of functions have factors $$\frac{1}{n!}$$. Mathlib doesn't do that because it has to work in every characteristic, but I think that `shift` was defined with that normalization in mind. So if it is to be the formal power series of the derivative, its definition has to be changed, preferably with a comment explaining why. (Problem: I don't know what that would break.)

So in the end, it is probably possible to do something using a modified version of `FormalMultilinearSeries.shift`, but I'm not sure that it would justify the extra effort, because I don't know what we would gain. We already know that, if `f` is continuously polynomial of degree `n`, then its derivative is continuously polynomial of degree `n-1`. We can get more explicit formulas for `f'(x)` in particular cases using the degree `1` term of the formal multilinear series of `f` at `x`; I did that for continuous multilinear maps, for example.

Reply:
Thanks. I looked into the definition of `shift` and it seems to be only currying the last coordinate, so it could only possibly work for a series whose terms are all symmetric multilinear forms. Even in that case, the nth term seems to be missing a factor of n. I think we should remove the claim that it's the derivative from the docstring.

Regarding the connection with linearDeriv, maybe the most general thing we can state and prove is that `linearDeriv` evaluate to the same thing as the `changeOriginSeries 1` of `toFormalMultilinearSeriesFixedDegree`. (Regarding the definition of `toFormalMultilinearSeriesFixedDegree`: I think you need to sum over powersetCard because you want a series development around an arbitrary point, but I think you can just define the series around the origin, which would just be a single term (a "homogeneous" CPolynomial), and then use changeOrigin of FormalMultilinearSeries rather than repeating the work.)

Reply:
I tracked down the history and it appears the claim that `FormalMultilinearSeries.shift` is the derivative was added in [mathlib3#2012](https://github.com/leanprover-community/mathlib/commit/4637e5c85ccd5c0798574d87005f94bf15b01bb3#diff-019a55ce3884d457d49a227ddffcf48744c88c3d5abe700250b85a8325d61ff1R183-R187) originally. Do you agree the claim is problematic, @sgouezel?

Reply:
I think the claim is correct: if for all `n` then `p (n+1)` is the derivative of `p n` (i.e., `p` is a Taylor series), then for all `n` one also has that `p (n+2)` is the derivative of `p (n+1)` (i.e., `p.shift` is a Taylor series, namely of the derivative of the original function). But there is indeed a catch, that for an analytic function its development as a power series does *not* coincide with its Taylor series (precisely of the `n!` factor). Let `f x = \sum q_n (x, ..., x)` be a power series, where `q_n` is `n `-multilinear. Then its `n`-th derivative at the origin is given by `D^n f (v_1, ..., v_n) = \sum_\sigma q_n (v_{\sigma(1), ..., v_{\sigma(n)})` where the sum is over all permutations of `n` symbols. It has to be something like that, because `D^n f` is symmetric while `q_n` has no reason to be symmetric. And this explains nicely the multiplicity `n!` that shows up in dimension `1`. Am I missing something here?

Reply:
Hi Sebastien, I don't know about Junyan, but for me the problem with that claim in mathlib is not that it cannot be made mathematically correct. I will believe you when ou say that it can. The problem is that I don't find it helpful, for two reasons:
(1) I don't know what the sentence "If `p` corresponds to the Taylor series of a function" means mathematically, and this is not explained in the file, so I was not able to translate the claim intro a precise mathematical sentence. (When Junyan asked me to compare `p.shift` with the `FormalMultilinearSeries` for the derivative of the sum of `p`, I actually went and googled "Taylor series", and I was not enlightened.)
(2) The claim is not proved anywhere.

Reply:
Your definition of Taylor series is weird and doesn't quite typecheck; surely (suppose we have a Taylor series in one variable) the derivative of $\frac{f^{(n)}(a)}{n!}x^n$ is $\frac{f^{(n)}(a)}{(n-1)!}x^{n-1}$, not $\frac{f^{(n+1)}(a)}{(n+1)!}x^{n+1}$?

To compute the derivative of the Taylor series we would like to go from $\frac{f^{(n)}(a)}{n!}x^n$ to $\frac{f^{(n)}(a)}{(n-1)!}x^{n-1}$. Assuming that the former is represented by the symmetric form $\frac{f^{(n)}(a)}{n!}x_1 x_2 \dots x_n$, we can set $x_n=1$ (derivative in $x$ is derivative in the direction $1$) and multiply by $n$ to get $\frac{f^{(n)}(a)}{(n-1)!}x_1 x_2 \dots x_{n-1}$, which evaluates to $\frac{f^{(n)}(a)}{(n-1)!}x^{n-1}$ as desired. However, this factor of $n$ is nowhere to be seen in the definition of `shift`. Also if it's not a symmetric form, you would omit each $x_i$ and sum the $n$ terms together, which is done in `changeOriginSeries 1`.

Let's ask the other author @urkud of [mathlib3#5878](https://github.com/leanprover-community/mathlib/commit/547d67fc7ad1ae9f0df0be5b64ac531c0dd5f4ca#diff-6090aa57561906639f8792a3240c2309c9e75b9900d9bb114e51f5cba12eab22R918-R925) to take a look.

Reply:
Sorry, I'm attributing to the wrong PR. The use of changeOriginSeries comes from [mathlib3#13127](https://github.com/leanprover-community/mathlib/pull/13127/files#diff-9a817f37e79c130ec39ff3455a5d5e282d63c9c86cee556db0183151eb46216fR77-R81) which was authored by @sgouezel alone.

Reply:
Okay, I looked up how `shift` is [used in mathlib](https://github.com/leanprover-community/mathlib4/blob/a59d0136f417dcbf1d9fceaa2d42ed87d592a164/Mathlib/Analysis/Calculus/ContDiff/Defs.lean#L560) and I think I now get where the confusion stems from. I think the "Taylor series" in the docstring refers to [HasFTaylorSeriesUpToOn](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/ContDiff/Defs.html#HasFTaylorSeriesUpToOn), where the $n$th term of the FormalMultilinearSeries is the "tensor of iterated/mixed $n$th-order partial derivatives". Evaluating this series doesn't give you back the original function (even if the function is analytic): the $n$th term needs to be divided by $n!$. So I'd say the docstring is true with an unconventional definition of "Taylor series". If we add a link to [`HasFTaylorSeriesUpToOn`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/ContDiff/Defs.html#HasFTaylorSeriesUpToOn) I think we can keep the docstring.

Reply:
And my other confusion (which led to my "doesn't typecheck" comment) comes from the difference between "derivatives" that we had in mind: in the setting of [HasFTaylorSeriesUpToOn](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/ContDiff/Defs.html#HasFTaylorSeriesUpToOn), the derivative of a function from `E` to `E^n →L[k] F` is an element of `E^{n+1} →L[k] F`. But in the setting of [HasFPowerSeriesOnBall](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Analytic/Basic.html#HasFPowerSeriesOnBall), the derivative of the evaluation of continuous multilinear function `E^n →L[k] F` at the diagonal is the evaluation of some continuous multilinear function `E^{n-1} →L[k] (E →L[k] F)` on the diagonal.

Reply:
I opened #9373 for the docs clarification.

Reply:
When I say "`f` has `p` as a Taylor series up to order `n`", I mean it in the sense of [HasFTaylorSeriesUpTo](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/ContDiff/Defs.html#HasFTaylorSeriesUpTo), i.e., `p 0` is equal to `f`, and `p (n+1)` is the derivative of `p n`. In this sense, a Taylor series is a gadget recording the sequence of derivatives of a function. The fact that `p.shift` is a Taylor series, in this sense, for the derivative of `f`, is for instance recorded in [hasFTaylorSeriesUpToOn_succ_iff_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/ContDiff/Defs.html#hasFTaylorSeriesUpToOn_succ_iff_right)

Reply:
@smorel394, could you have a look at #9373 to see if you think the clarification of the docs is good enough for you, or if you think you can make it even better for future users?

Reply:
It looks good to me.

Reply:
> Regarding the connection with linearDeriv, maybe the most general thing we can state and prove is that `linearDeriv` evaluate to the same thing as the `changeOriginSeries 1` of `toFormalMultilinearSeriesFixedDegree`. (Regarding the definition of `toFormalMultilinearSeriesFixedDegree`: I think you need to sum over powersetCard because you want a series development around an arbitrary point, but I think you can just define the series around the origin, which would just be a single term (a "homogeneous" CPolynomial), and then use changeOrigin of FormalMultilinearSeries rather than repeating the work.)

This is really about PR #9130, I'm making changes there to answer your suggestion.

Reply:
Done, I've also rewritten the proof that continuous multilinear maps are analytic in [another project](https://github.com/smorel394/ExteriorPowers/blob/master/ExteriorPowers/ContinuousMultilinearMapContDiff.lean). The applications to differentiability require completeness right now, but only until this PR is merged.

I got rid of the formal multilinear series stuff in #9130, since the definition of the formal multilinear series at 0 is so basic and can be done very quickly when needed. The proof that the `fderiv` is equal to the degree 1 of `changeOrigin` of the formal multilinear series at 0 is a bit messy for now, but at least it compiles, and I'm sure I can make it shorter.